### PR TITLE
Address edge case stream allocation

### PR DESCRIPTION
**Pull request by boks1971**
(NOTE: Will drop more notes inline)

A very edge case
- Available layer: [0, 1, 2], but bitrate is not yet available.
We set it to layer 2 awaiting measurement.
- Measurement for layers 0 and 1 come through.
- Still no key frame for layer 2.
- Finalize layers runs and sees that bitrate is available for 0 and 1.
It finalizes layer 1.
- Layer 1 key frame comes (because we asked key frame of layer 2,
publisher sends key frame for all layers). Locks to layer 1.
- No more events happen to switch to layer 2.

Changes
-------
- Move bit rate measurement to StreamTrackerManager. Allows re-use
in relay.
- Make bit rate availability (from zero -> non-zero) an event
and let it flow through the stream allocation flow so that we
always have an event when bit rate measurement becomes available.
This gets rid of finalization which I was unhappy with it anyway as
it was polling every second.
- Removing REMB stuff from buffer. We do not use it.
It is incorrect anyway. REMB should be ay peer connection level.

### CODE REVIEW

File: pkg/sfu/buffer/buffer.go [hunk 1]
```diff
@@ -18,22 +18,20 @@ import (
...
-	Payload   interface{}
-	KeyFrame  bool
-	RawPacket []byte
+	Head          bool
+	Arrival       int64
+	Packet        *rtp.Packet
+	Payload       interface{}
+	KeyFrame      bool
+	RawPacket     []byte
+	SpatialLayer  int32
```
Comment (boks1971) [inline pkg/sfu/buffer/buffer.go:33]: Recording spatial layer and temporal layer in `ExtPacket` which gets forwarded. `SpatialLayer` is forward looking for SVC style codecs. There will be a single buffer (i. e. single SSRC) which will have spatial layer information also.

NOTE: Thinking about including temporal layer in relay protocol, but not decided yet.

File: pkg/sfu/buffer/buffer.go [hunk 2]
```diff
@@ -66,14 +62,11 @@ type Buffer struct {
 	latestTSForAudioLevel            uint32
 
 	lastPacketRead int
-	bitrate        atomic.Value
-	bitrateHelper  [4]int64
```
Comment (boks1971) [inline pkg/sfu/buffer/buffer.go:70]: All the bitrate calculation moves to `StreamTrackerManager` which can be re-used in relay also.

File: pkg/sfu/buffer/buffer.go [hunk 3]
```diff
@@ -159,6 +149,7 @@ func (b *Buffer) Bind(params webrtc.RTPParameters, codec webrtc.RTPCodecCapabili
 			switch fb.Type {
 			case webrtc.TypeRTCPFBGoogREMB:
 				b.logger.Debugw("Setting feedback", "type", webrtc.TypeRTCPFBGoogREMB)
+				b.logger.Warnw("REMB not supported, RTCP feedback will not be generated", nil)
```
Comment (boks1971) [inline pkg/sfu/buffer/buffer.go:152]: Warning about REMB here (it will print every time, but that is okay I think). The REMB building here is unused. It was incorrect anyway because REMB should be aggregated at peer connection level.

File: pkg/sfu/downtrack.go [hunk 4]
```diff
@@ -643,6 +647,14 @@ func (d *DownTrack) UpTrackLayersChange(availableLayers []int32) {
 	}
 }
 
+func (d *DownTrack) UpTrackBitrateAvailabilityChange() {
```
Comment (boks1971) [inline pkg/sfu/downtrack.go:650]: New notification when bitrate calculation is available.

File: pkg/sfu/forwarder.go [hunk 5]
```diff
@@ -311,18 +311,38 @@ func (f *Forwarder) UpTrackLayersChange(availableLayers []int32) {
...
+
+		for t := f.maxLayers.temporal; t >= 0; t-- {
+			if brs[layer][t] != 0 {
+				bitrateAvailableLayers = append(bitrateAvailableLayers, layer)
+				break
+			}
+		}
+	}
+
+	return optimalBandwidthNeeded, len(f.availableLayers) == len(bitrateAvailableLayers)
```
Comment (boks1971) [inline pkg/sfu/forwarder.go:345]: This is enhanced to check if all available layers have bitrate available.

File: pkg/sfu/receiver.go [hunk 6]
```diff
@@ -498,8 +495,8 @@ func (w *WebRTCReceiver) forwardRTP(layer int32) {
 			return
 		}
 
-		if tracker != nil {
-			tracker.Observe(pkt.Packet.SequenceNumber)
+		if tracker != nil && len(pkt.Packet.Payload) > 0 {
```
Comment (boks1971) [inline pkg/sfu/receiver.go:498]: Dropping padding only packets from stream tracker.

File: pkg/sfu/streamtracker.go [hunk 7]
```diff
@@ -44,27 +48,32 @@ type StreamTrackerParams struct {
 type StreamTracker struct {
 	params StreamTrackerParams
 
-	onStatusChanged func(status StreamStatus)
+	onStatusChanged    func(status StreamStatus)
+	onBitrateAvailable func()
+
+	lock sync.RWMutex
```
Comment (boks1971) [inline pkg/sfu/streamtracker.go:54]: Using lock here consistently. There were too many atomics and it was getting hard to follow. All the functions here are very well scoped and extremely small. So, just use the lock consistently.

File: pkg/sfu/streamtracker.go [hunk 8]
```diff
@@ -133,90 +148,175 @@ func (s *StreamTracker) Stop() {
...
 }
 
 func (s *StreamTracker) SetPaused(paused bool) {
-	s.paused.Store(paused)
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.paused = paused
+
+	if !paused {
```
Comment (boks1971) [inline pkg/sfu/streamtracker.go:186]: Let `pause` declare the layers inactive and resetting coming out of pause (similar behaviour to dynacast when we expect a layer to start up) for a quick start. Keep things cleaner.

File: pkg/sfu/buffer/buffer.go (hunk 1 above)
Comment (davidzhao) [inline pkg/sfu/buffer/buffer.go:33]: seems that putting these in ExtPacket will make it easier for us to support AV1/VP9 and the likes?

File: pkg/sfu/forwarder.go [hunk 9]
```diff
@@ -396,43 +416,41 @@ func (f *Forwarder) Allocate(availableChannelCapacity int64, allowPause bool, br
...
-					// and if it will fit in the available channel capacity
-					state = VideoAllocationStateDeficient
+				// disable forwarding as it is not known how big this stream is
+				// and if it will fit in the available channel capacity
+				state = VideoAllocationStateDeficient
 
-					if f.targetLayers != InvalidLayers {
-						change = VideoStreamingChangePausing
-					}
+				if f.targetLayers != InvalidLayers {
```
Comment (davidzhao) [inline pkg/sfu/forwarder.go:452]: Can we trust our BWE enough to pause? I'm concerned it'll pause unnecessarily due to inaccurate BWE prints.

Seems better to lean on the side of non-congestion since it's more likely.. wdyt?

**Review by davidzhao — APPROVED** [review]
I actually saw this behavior during testing and wanted to get enough details to file a bug. You are one step ahead of me as usual! 

One piece of feedback about pausing when stream bandwidth is unknown.

File: pkg/sfu/forwarder.go (hunk 9 above)
Comment (boks1971) [reply pkg/sfu/forwarder.go:452]: Yeah. Going to re-do this function as this is used in only one scenario, i. e. allocate the best available layer. Will clean up this function so that it is not confusing.

File: pkg/sfu/buffer/buffer.go (hunk 1 above)
Comment (boks1971) [reply pkg/sfu/buffer/buffer.go:37]: Yes, that is the intention.

**Comment by boks1971** [comment]
@davidzhao Ready for a re-review. Cleaned up/simplified bits. How much to rely on BWE and soft handling those may require more data/discussions. This PR is not necessarily changing how we respond to BWE in any significant way.

**Review by cnderrauber — APPROVED** [review]
